### PR TITLE
8248404: AArch64: Remove uses of long and unsigned long

### PR DESCRIPTION
**Pull request by coleenp**
Tested with mach5 on linux-aarch64 and macosx-aarch64 on tier1-3 and below GHA for windows-aarch64 (once I open this PR).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8248404](https://bugs.openjdk.java.net/browse/JDK-8248404): AArch64: Remove uses of long and unsigned long

### Reviewers
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to 232e1d635a8cd5cc529f26ef3cbb05e3f33db229

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7023/head:pull/7023` \
`$ git checkout pull/7023`

Update a local copy of the PR: \
`$ git checkout pull/7023` \
`$ git pull https://git.openjdk.java.net/jdk pull/7023/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7023`

View PR using the GUI difftool: \
`$ git pr show -t 7023`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7023.diff">https://git.openjdk.java.net/jdk/pull/7023.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64_log.cpp [hunk 1]
```diff
@@ -297,7 +297,8 @@ void MacroAssembler::fast_log(FloatRegister vtmp0, FloatRegister vtmp1,
   bind(MAIN);
     fmovs(tmp3, vtmp5);                        // int intB0 = AS_INT_BITS(B);
     mov(tmp5, 0x3FE0);
-    mov(rscratch1, 0xffffe00000000000);
+    uint64_t mask = 0xffffe00000000000UL;
```
Comment (stefank) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64_log.cpp:300]: I think this should be using ULL to support LLP64.

**Review by stefank — CHANGES_REQUESTED** [review]

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp [hunk 2]
```diff
@@ -492,18 +492,16 @@ class MacroAssembler: public Assembler {
...
 
   inline void mov(Register dst, address addr)             { mov_immediate64(dst, (uint64_t)addr); }
+  inline void mov(Register dst, int32_t imm32)            { mov_immediate64(dst, imm32); }
+  inline void mov(Register dst, uint32_t imm32)           { mov_immediate64(dst, imm32); }
+#ifdef __APPLE__
+  // macosx wants all the overloads
+  inline void mov(Register dst, intptr_t imm32)           { mov_immediate64(dst, imm32); }
+#endif
+  inline void mov(Register dst, int64_t imm32)           { mov_immediate64(dst, imm32); }
+  inline void mov(Register dst, uint64_t imm32)           { mov_immediate64(dst, imm32); }
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp:502]:
```suggestion
  inline void mov(Register dst, uint64_t imm64)           { mov_immediate64(dst, imm64); }
```

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp [hunk 3]
```diff
@@ -492,18 +492,16 @@ class MacroAssembler: public Assembler {
   // 64 bit integers
 
   inline void mov(Register dst, address addr)             { mov_immediate64(dst, (uint64_t)addr); }
+  inline void mov(Register dst, int32_t imm32)            { mov_immediate64(dst, imm32); }
+  inline void mov(Register dst, uint32_t imm32)           { mov_immediate64(dst, imm32); }
+#ifdef __APPLE__
+  // macosx wants all the overloads
+  inline void mov(Register dst, intptr_t imm32)           { mov_immediate64(dst, imm32); }
+#endif
+  inline void mov(Register dst, int64_t imm32)           { mov_immediate64(dst, imm32); }
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp:501]:
```suggestion
  inline void mov(Register dst, int64_t imm64)           { mov_immediate64(dst, imm64); }
```

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp [hunk 4]
```diff
@@ -492,18 +492,16 @@ class MacroAssembler: public Assembler {
   // 64 bit integers
 
   inline void mov(Register dst, address addr)             { mov_immediate64(dst, (uint64_t)addr); }
+  inline void mov(Register dst, int32_t imm32)            { mov_immediate64(dst, imm32); }
+  inline void mov(Register dst, uint32_t imm32)           { mov_immediate64(dst, imm32); }
+#ifdef __APPLE__
+  // macosx wants all the overloads
+  inline void mov(Register dst, intptr_t imm32)           { mov_immediate64(dst, imm32); }
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp:499]:
```suggestion
  inline void mov(Register dst, intptr_t imm64)           { mov_immediate64(dst, imm64); }
```

File: src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp [hunk 5]
```diff
@@ -71,7 +71,7 @@ address ProgrammableUpcallHandler::generate_upcall_stub(jobject rec, jobject jab
 
   // Capture prev stack pointer (stack arguments base)
   __ add(rscratch1, rfp, 16);   // Skip saved FP and LR
-  __ str(rscratch1, Address(sp, layout.stack_args));
+  __ str(rscratch1, Address(sp, (int)layout.stack_args));  // x86 casts to int also
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp:74]:
```suggestion
  __ str(rscratch1, Address(sp, checked_cast<int>(layout.stack_args)));  // x86 casts to int also
```

Comment (theRealAph) [reply src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp:74]: ... because this is UB if it doesn't fit.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64_log.cpp (hunk 1 above)
Comment (merykitty) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64_log.cpp:300]: Integral literal takes the first type that fits so for this 64-bit all U, UL and ULL would produce the same type.

Comment (stefank) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64_log.cpp:300]: Thanks, I wasn't sure what would happen if you used the "wrong" suffix. I think it still makes sense to use the correct prefix for the intended target type.

Comment (coleenp) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64_log.cpp]: I changed it to ULL. Thanks.

File: src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp (hunk 5 above)
Comment (coleenp) [reply src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp]: Yes, I changed this and wasn't really sure how to fix it better.  I'd love if Address had a unit64_t argument but all the callers that just passed 0 wouldn't compile.  The code is like that in the x86 version. So I thought that'd be safe.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp (hunk 4 above)
Comment (coleenp) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp]: fixed.

File: src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp (hunk 5 above)
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp]: Well, I'm not convinced the x86 version is safe, really. It probably is, but the checked_cast costs nothing in production.

Comment (theRealAph) [reply src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp]: Actually, thinking some more, I can see no guarantee that this `Address` is valid either. So I'm going to change my mind.

File: src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp [hunk 6]
```diff
@@ -71,7 +71,7 @@ address ProgrammableUpcallHandler::generate_upcall_stub(jobject rec, jobject jab
 
   // Capture prev stack pointer (stack arguments base)
   __ add(rscratch1, rfp, 16);   // Skip saved FP and LR
-  __ str(rscratch1, Address(sp, layout.stack_args));
+  __ str(rscratch1, Address(sp, checked_cast<int>(layout.stack_args)));  // x86 casts to int also
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp:74]:
```suggestion
  __ Address slot = __ legitimize_address(Address(sp, checked_cast<int>(layout.stack_args)), wordSize, rscratch2);
  __ str(rscratch1, slot);  // x86 casts to int also
```
I think this is a real bug: the range of a stack arg from SP can exceed that of the maximum offset of a STR instruction!
Wherever there's a dubious cast there's probably a bug...

Comment (theRealAph) [reply src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp:74]: All of this may seem tedious and pedantic, but we have had failures in production caused by stack pointer offsets exceeding the 12-bit range of a STR instruction.

Comment (coleenp) [reply src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp:74]: Yes, I agree and didn't really know how to fix it so it would compile (if it actually compiles now) and not be UB.  I'm close to giving up!  Thanks for the code change.

Comment (coleenp) [reply src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp:74]: It does seem pedantic which makes it really difficult, but it's important to get this right.  These kinds of bugs are horrible to debug.

File: src/hotspot/cpu/aarch64/assembler_aarch64.hpp [hunk 7]
```diff
@@ -406,16 +406,6 @@ class Address {
...
-  Address(Register r, long o)
-    : _base(r), _index(noreg), _offset(o), _mode(base_plus_offset), _target(0) { }
-  Address(Register r, long long o)
-    : _base(r), _index(noreg), _offset(o), _mode(base_plus_offset), _target(0) { }
-  Address(Register r, unsigned int o)
-    : _base(r), _index(noreg), _offset(o), _mode(base_plus_offset), _target(0) { }
-  Address(Register r, unsigned long o)
-    : _base(r), _index(noreg), _offset(o), _mode(base_plus_offset), _target(0) { }
-  Address(Register r, unsigned long long o)
-    : _base(r), _index(noreg), _offset(o), _mode(base_plus_offset), _target(0) { }
```
[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/cpu/aarch64/assembler_aarch64.hpp:418]: Will it silently truncate?  I was getting warnins.  with the unit64_t overloads, I was getting ambiguous calls with address or Register.  Maybe that's equally wrong.

[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/cpu/aarch64/assembler_aarch64.hpp:418]: Getting closer to giving up ...

[the remark this answers is not in the document]
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/assembler_aarch64.hpp:418]: It seems to silently truncate if the only signed type is int. I just put a breakpoint on `LIR_Assembler::as_Address` and watched it happen.

[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/cpu/aarch64/assembler_aarch64.hpp:418]: There are too many overloads. I don't know why long long overloads didn't cause ambiguity but if there's any conversion needed, it's ambiguous.
Address's _offset field is an int64_t, which seems wrong to assign a unint64_t into it.

[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/cpu/aarch64/assembler_aarch64.hpp:418]: I added a ptrdiff_t and a unint64_t overload.  The linux build is happy about that.  I'm going to push so the GHA tries the rest.

[the remark this answers is not in the document]
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/assembler_aarch64.hpp:418]: I just got a clean build with overloads for `int`,` int64_t`, and `uint64_t` offsets. That seems to be the minimum.

File: src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp (hunk 5 above)
Comment (coleenp) [reply src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp]: I'm going to remove the cast or cast it to uint64_t in universalUpcallHandler. thanks for bearing with me!

File: src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp (hunk 6 above)
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp]: OK, so if we have the `Address(reg_offset)` overloads for `int`, `int64_t`, and uint64_t` we're good, and by calling the right overload we get rid of the need to cast the `layout.stack_args` offset to `int` here.

Comment (theRealAph) [reply src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp]: This is what I've got:
```

  Address(Register r)
    : _base(r), _index(noreg), _offset(0), _mode(base_plus_offset), _target(0) { }
  Address(Register r, int o)
    : _base(r), _index(noreg), _offset(o), _mode(base_plus_offset), _target(0) { }
  Address(Register r, int64_t o)
    : _base(r), _index(noreg), _offset(o), _mode(base_plus_offset), _target(0) { }
  Address(Register r, uint64_t o)
    : _base(r), _index(noreg), _offset(o), _mode(base_plus_offset), _target(0) { }
  Address(Register r, ByteSize disp)
    : Address(r, in_bytes(disp)) { }

```

Comment (coleenp) [reply src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp]: I added ptrdiff_t but that's the same as int64_t.  I'll fix it.

File: src/hotspot/cpu/aarch64/assembler_aarch64.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/cpu/aarch64/assembler_aarch64.hpp:418]: what platforms did you get this to compile on?  I only have a linux and the macosx compiler seems to have different rules as well as windows, which only have with GHA.

[the remark this answers is not in the document]
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/assembler_aarch64.hpp:418]: Sounds good.

[the remark this answers is not in the document]
Comment (merykitty) [reply src/hotspot/cpu/aarch64/assembler_aarch64.hpp:418]: Integral types are implicitly convertible to each other, I believe in this case a single overload for `int64_t` would be sufficient?

[the remark this answers is not in the document]
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/assembler_aarch64.hpp:418]: On Linux. The MacOS debugger makes me pull my hair out.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64_log.cpp (hunk 1 above)
Comment (kimbarrett) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64_log.cpp]: We have `CONST64` and `UCONST64` for these situations (in globalDefinitions.hpp).

Comment (coleenp) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64_log.cpp]: Forgot about those. thanks.

File: src/hotspot/cpu/aarch64/assembler_aarch64.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/cpu/aarch64/assembler_aarch64.hpp:418]: I think I tried int64_t instead of plain 'int' and got errors for simple Address(reg, 0).

[the remark this answers is not in the document]
Comment (merykitty) [reply src/hotspot/cpu/aarch64/assembler_aarch64.hpp:418]: Oh right I now realize that integral promotion only promotes things to `int`. So it get ambiguous cause both `int -> int64_t` and `int -> ByteSize` are integral conversion. Ouch.

File: src/hotspot/cpu/aarch64/assembler_aarch64.hpp [hunk 8]
```diff
@@ -406,15 +406,16 @@ class Address {
...
     : _base(r), _index(noreg), _offset(o), _mode(base_plus_offset), _target(0) { }
-  Address(Register r, unsigned int o)
+  Address(Register r, ptrdiff_t o)
     : _base(r), _index(noreg), _offset(o), _mode(base_plus_offset), _target(0) { }
-  Address(Register r, unsigned long o)
-    : _base(r), _index(noreg), _offset(o), _mode(base_plus_offset), _target(0) { }
-  Address(Register r, unsigned long long o)
+#endif
+  Address(Register r, uint64_t o)
     : _base(r), _index(noreg), _offset(o), _mode(base_plus_offset), _target(0) { }
```
Comment (phedlin) [inline src/hotspot/cpu/aarch64/assembler_aarch64.hpp:419]: The above could be written (from an overlapping CS, with additional initialisation)
```suggestion
    Address(Register r) : Address(r, 0) {}
    Address(Register r, int32_t o) : Address(r, int64_t(o)) {}
    Address(Register r, int64_t o) : _mode(base_plus_offset),
      _base(r), _index(noreg), _offset(o), _extend(lsl(0)), _target(nullptr) {}

    Address(Register r, uint32_t o) : Address(r, uint64_t(o)) {}
    Address(Register r, uint64_t o) : _mode(base_plus_offset),
      _base(r), _index(noreg), _offset(o), _extend(lsl(0)), _target(nullptr) {}
```

File: src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp (hunk 6 above)
Comment (coleenp) [reply src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp]: I need ptrdiff_t additionally to get this to compile on our macosx-aarch64 platform.

File: src/hotspot/cpu/aarch64/assembler_aarch64.hpp [hunk 9]
```diff
@@ -404,18 +405,11 @@ class Address {
...
-    : _base(r), _index(noreg), _offset(o), _mode(base_plus_offset), _target(0) { }
-  Address(Register r, unsigned long o)
-    : _base(r), _index(noreg), _offset(o), _mode(base_plus_offset), _target(0) { }
-  Address(Register r, unsigned long long o)
-    : _base(r), _index(noreg), _offset(o), _mode(base_plus_offset), _target(0) { }
+
+  template<typename T, ENABLE_IF(std::is_integral<T>::value)>
+  Address(Register r, T o)
+    : _base(r), _index(noreg), _offset(o), _mode(base_plus_offset), _target(0) {}
+
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/assembler_aarch64.hpp:412]: Hum, interesting. Looks good: all integer types should get sign- or zero-extended as appropriate, I think. I am going to check tho'. :-)

Comment (coleenp) [reply src/hotspot/cpu/aarch64/assembler_aarch64.hpp:412]: Great, thanks for checking!

Comment (theRealAph) [reply src/hotspot/cpu/aarch64/assembler_aarch64.hpp:412]: Seems to be fine: does the sign extension right, doesn't do any bogus translation. Ship it!

Comment (coleenp) [reply src/hotspot/cpu/aarch64/assembler_aarch64.hpp:412]: Thanks Andrew.  Now I need another reviewer and you to approve. Thanks for all your help.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp [hunk 10]
```diff
@@ -493,17 +494,10 @@ class MacroAssembler: public Assembler {
...
   inline void mov(Register dst, address addr)             { mov_immediate64(dst, (uint64_t)addr); }
 
-  inline void mov(Register dst, int imm64)                { mov_immediate64(dst, (uint64_t)imm64); }
-  inline void mov(Register dst, long imm64)               { mov_immediate64(dst, (uint64_t)imm64); }
-  inline void mov(Register dst, long long imm64)          { mov_immediate64(dst, (uint64_t)imm64); }
-  inline void mov(Register dst, unsigned int imm64)       { mov_immediate64(dst, (uint64_t)imm64); }
-  inline void mov(Register dst, unsigned long imm64)      { mov_immediate64(dst, (uint64_t)imm64); }
-  inline void mov(Register dst, unsigned long long imm64) { mov_immediate64(dst, (uint64_t)imm64); }
+  template<typename T, ENABLE_IF(std::is_integral<T>::value)>
+  inline void mov(Register dst, T o)                      { mov_immediate64(dst, (uint64_t)o); }
```
Comment (kimbarrett) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp:498]: I'm not sure why this isn't just
`inline void mov(Register dst, uint64_t o) { mov_immediate64(dst, o); }`
Unlike the `Address` case, I don't see any ambiguous implicit conversions here.

File: src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp [hunk 11]
```diff
@@ -1171,7 +1171,7 @@ static void verify_oop_args(MacroAssembler* masm,
         VMReg r = regs[i].first();
         assert(r->is_valid(), "bad oop arg");
         if (r->is_stack()) {
-          __ ldr(temp_reg, Address(sp, r->reg2stack() * VMRegImpl::stack_slot_size));
+          __ ldr(temp_reg, Address(sp, (uint64_t)r->reg2stack() * VMRegImpl::stack_slot_size));
```
Comment (kimbarrett) [inline src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp:1174]: Why is the cast being added here?  Is it because the multiply can overflow an int, and this is really fix for a bug that is distinct from the cleanup in this PR?  There are a couple more like this later in this file.

File: src/hotspot/cpu/aarch64/templateInterpreterGenerator_aarch64.cpp [hunk 12]
```diff
@@ -969,7 +969,7 @@ address TemplateInterpreterGenerator::generate_CRC32_update_entry() {
 /**
  * Method entry for static native methods:
  *   int java.util.zip.CRC32.updateBytes(int crc, byte[] b, int off, int len)
- *   int java.util.zip.CRC32.updateByteBuffer(int crc, long buf, int off, int len)
+ *   int java.util.zip.CRC32.updateByteBuffer(int crc, jlong buf, int off, int len)
```
Comment (kimbarrett) [inline src/hotspot/cpu/aarch64/templateInterpreterGenerator_aarch64.cpp:972]: Why not `int64_t` instead of `jlong`?   Similarly in a couple later places.

File: src/hotspot/cpu/aarch64/templateInterpreterGenerator_aarch64.cpp [hunk 13]
```diff
@@ -994,7 +994,7 @@ address TemplateInterpreterGenerator::generate_CRC32_updateBytes_entry(AbstractI
     // Arguments are reversed on java expression stack
     // Calculate address of start element
     if (kind == Interpreter::java_util_zip_CRC32_updateByteBuffer) {
-      __ ldr(buf, Address(esp, 2*wordSize)); // long buf
+      __ ldr(buf, Address(esp, 2*wordSize)); // jlong buf
```
Comment (kimbarrett) [inline src/hotspot/cpu/aarch64/templateInterpreterGenerator_aarch64.cpp:997]: Maybe just drop the type in the comment?  Similarly in a couple later places.

File: src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp [hunk 14]
```diff
@@ -71,7 +71,8 @@ address ProgrammableUpcallHandler::generate_upcall_stub(jobject rec, jobject jab
 
   // Capture prev stack pointer (stack arguments base)
   __ add(rscratch1, rfp, 16);   // Skip saved FP and LR
-  __ str(rscratch1, Address(sp, layout.stack_args));
+  Address slot = __ legitimize_address(Address(sp, layout.stack_args), wordSize, rscratch2);
```
Comment (kimbarrett) [inline src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp:74]: Introduction of legitimize_address here seems unrelated to this PR.  What's this about?

**Review by kimbarrett — CHANGES_REQUESTED** [review]

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp (hunk 10 above)
Comment (coleenp) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp:498]: Because the macosx compiler complains of ambiguity for the mov also if this is the only mov.

File: src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp (hunk 11 above)
Comment (coleenp) [reply src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp:1174]: This was here to match one of the Address constructors that I had.  With the template, it's no longer needed.

File: src/hotspot/cpu/aarch64/templateInterpreterGenerator_aarch64.cpp (hunk 12 above)
Comment (coleenp) [reply src/hotspot/cpu/aarch64/templateInterpreterGenerator_aarch64.cpp:972]: This comment refers to Java code which more appropriately uses jlong.  Actually, the Java code uses long.  I'll revert this change.

File: src/hotspot/cpu/aarch64/templateInterpreterGenerator_aarch64.cpp (hunk 13 above)
Comment (coleenp) [reply src/hotspot/cpu/aarch64/templateInterpreterGenerator_aarch64.cpp:997]: Ok. That'll give me less 'long' matches.

Comment (coleenp) [reply src/hotspot/cpu/aarch64/templateInterpreterGenerator_aarch64.cpp:997]: No, it's useful because it describes the parameter that you're loading.  I'm leaving this jlong.

File: src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp (hunk 14 above)
Comment (coleenp) [reply src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp:74]: This is a bug that Andrew pointed out. I could file a different CR for this and take it out.  It seemed a minor thing to include with this change though.

**Review by theRealAph — APPROVED** [review]

File: src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp (hunk 14 above)
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp:74]: > Introduction of legitimize_address here seems unrelated to this PR. What's this about?

SP-relative loads and stores have only a 12-bit offset. This is a bug that was discovered during review.

**Comment by coleenp** [comment]
Thanks @theRealAph for your help and the review.

File: src/hotspot/cpu/aarch64/templateInterpreterGenerator_aarch64.cpp (hunk 13 above)
Comment (coleenp) [reply src/hotspot/cpu/aarch64/templateInterpreterGenerator_aarch64.cpp:997]: Or I'll revert this whole file.

File: src/hotspot/cpu/aarch64/templateInterpreterGenerator_aarch64.cpp (hunk 12 above)
Comment (kimbarrett) [reply src/hotspot/cpu/aarch64/templateInterpreterGenerator_aarch64.cpp]: Yeah, this is in a comment referring to a Java method signature, so I agree `long` is correct here.

File: src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp (hunk 14 above)
Comment (kimbarrett) [reply src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp:74]: Thanks for the explanation.

**Review by kimbarrett — APPROVED** [review]
Looks good.

File: src/hotspot/cpu/aarch64/templateInterpreterGenerator_aarch64.cpp [hunk 15]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2003, 2021, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2003, 2022, Oracle and/or its affiliates. All rights reserved.
```
Comment (dholmes-ora) [inline src/hotspot/cpu/aarch64/templateInterpreterGenerator_aarch64.cpp:2]: No other changes to this file, so this needs reverting.

**Review by dholmes-ora — COMMENTED** [review]
Just a drive-by comment below.

Cheers,
David

File: src/hotspot/cpu/aarch64/templateInterpreterGenerator_aarch64.cpp (hunk 15 above)
Comment (coleenp) [reply src/hotspot/cpu/aarch64/templateInterpreterGenerator_aarch64.cpp:2]: Oh, ok, thanks.

File: src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp (hunk 11 above)
Comment (coleenp) [reply src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp]: I think I didn't revert this.  Retesting.

**Comment by coleenp** [comment]
There were a couple of files changed that still had casts for the intermediate changes before templates.  I reverted them and reran tier1 testing.  I'll push after GHA completes.

**Review by theRealAph — COMMENTED** [review]
Even better! I still approve.

**Comment by coleenp** [comment]
Thank you for the help and reviews!
/integrate
